### PR TITLE
Improve conhost's scrolling performance

### PR DESCRIPTION
**Pull request by lhecker**
`EnableScrollbar()` and especially `SetScrollInfo()` are prohibitively
expensive functions nowadays. This improves throughput of good old
`type` in cmd.exe by ~10x, by briefly releasing the console lock.

## Validation Steps Performed
* `type`ing a file in `cmd` is as fast while the window is scrolling
  as it is while it isn't scrolling ✅
* Scrollbar pops in and out when scroll-forward is disabled ✅

### CODE REVIEW

**Comment by lhecker** [comment]
Before:

https://github.com/microsoft/terminal/assets/2256941/762196b0-16d3-4768-b29d-26ff92abfde1

After:

https://github.com/microsoft/terminal/assets/2256941/6ceeecf9-3525-43bf-8238-203e076cc063

**Comment by DHowett** [comment]
_what on earth sorcery is this_

File: src/interactivity/onecore/ConsoleWindow.hpp [hunk 1]
```diff
@@ -24,8 +24,7 @@ namespace Microsoft::Console::Interactivity::OneCore
     {
     public:
         // Inherited via IConsoleWindow
-        BOOL EnableBothScrollBars() noexcept override;
-        int UpdateScrollBar(bool isVertical, bool isAltBuffer, UINT pageSize, int maxSize, int viewportPosition) noexcept override;
+        int UpdateScrollBar(bool isAltBuffer, til::size maxSize, const til::rect& viewport) noexcept override;
```
Comment (DHowett) [inline src/interactivity/onecore/ConsoleWindow.hpp:27]: this one says `bar`, others say `bars`.

**Review by DHowett — DISMISSED** [review]

**Comment by zadjii-msft** [comment]
holy fuck what

File: src/host/renderData.cpp [hunk 2]
```diff
@@ -87,14 +87,16 @@ std::vector<Viewport> RenderData::GetSelectionRects() noexcept
...
-    ::LockConsole();
+    auto& gci = ServiceLocator::LocateGlobals().getConsoleInformation();
+    gci.LockConsole();
 }
 
 // Method Description:
 // - Unlocks the console after a call to RenderData::LockConsole.
 void RenderData::UnlockConsole() noexcept
 {
-    ::UnlockConsole();
```
[the remark this answers is not in the document]
Comment (lhecker) [reply src/host/renderData.cpp:97]: Yes, this isn't the right place to emit the CTRL event IMO. There are only 3 situations where calling the global `UnlockConsole` makes sense: The console IO thread (and technically only when handling `GenerateConsoleCtrlEvent`), when handling manual input (because of Ctrl+C) or during shutdown.

As far as I can see, `::UnlockConsole` and it calling `ProcessCtrlEvents()` is just a hold-over from v0.5 code which didn't quite get perfectly integrated into v1 and then into v2, because originally only the global `UnlockConsole` function existed and the one on `CONSOLE_INFORMATION` is a much newer C++ addition. Putting it into `::UnlockConsole` initially makes sense IMO, since the check is cheap and it probably reduced binary size.

But if you think about it, it's not frightening at all and it's also not something we need to keep. That's because we _know_ exactly when a Ctrl event was added: Only and exclusively through `HandleCtrlEvent`. That function is only called from 3 places: The `GenerateConsoleCtrlEvent` console API handler, the conhost message handler `HandleKeyEvent` or in response to closing the conhost window via `CloseConsoleProcessState`.

(Side note: I'm now confused why ConPTY doesn't call `CloseConsoleProcessState`. That's probably not good...)

[the remark this answers is not in the document]
Comment (DHowett) [reply src/host/renderData.cpp:97]: > That function is only called from 3 places

What about when we insert a <kbd>Ctrl+C</kbd> into the input buffer via InputStateMachineEngine?

[the remark this answers is not in the document]
Comment (lhecker) [reply src/host/renderData.cpp:97]: That goes through `HandleKeyEvent`'s off-spring `HandleGenericKeyEvent` (which was split off from it).

**Review by zadjii-msft — DISMISSED** [review]

File: src/host/screenInfo.hpp [hunk 3]
```diff
@@ -99,8 +99,14 @@ class SCREEN_INFORMATION : public ConsoleObjectHeader, public Microsoft::Console
...
 
+    struct ScrollBarState
+    {
+        til::size maxSize;
+        til::rect viewport;
+        bool isAltBuffer = false;
+    };
     void UpdateScrollBars();
-    void InternalUpdateScrollBars();
+    ScrollBarState FetchScrollBarState();
```
Comment (lhecker) [inline src/host/screenInfo.hpp:109]: The problem with the tests was the briefly unlocking the console allowed window-resize messages to creep in.

To fix that, I had to do what I explained last week in our meeting: Instead of reaching through `SCREEN_INFORMATION` back into `Window`, I now leave control entirely inside `Window`. This allows me to properly unlock the console at the right time and also leave it unlocked.
This however made it necessary to move `SCREEN_INFORMATION::ResizingWindow` into `Window`, since the console lock is now not being held anymore. That's also why all the calls got changed from `InternalUpdateScrollBars` to just `UpdateScrollBars`. That is, now we're updating the scrollbars asynchronously during launch, etc. In my testing this works just fine.

This change is also another example for why I'm convinced that "loops" in control flow should be avoided (unless they're a good fit of course). Here, `Window` calls into `SCREEN_INFORMATION` and it calls back into `Window` with some extra data. That is very similar to our other callback-centric designs. I don't believe it's a coincidence that they had the same problems too. For instance `TerminalInput` accessing `Terminal` state without holding locks, is approximately similar to this code holding the console lock despite not needing it.
Basically I'm advocating for the latter in this graph, and I believe most of the former designs can be transformed to the latter, while retaining all the benefits:
```mermaid
sequenceDiagram
    participant A as class A
    participant B as class B
    participant C as class C
    opt A/B share control
    activate A
    A->>B: foo()
    activate B
    B->>C: bar()
    C-->>B: 
    B-->>A: 
    deactivate B
    deactivate A
    end
    opt only A is in control
    activate A
    A->>B: foo()
    B-->>A: 
    deactivate A
    activate A
    A->>C: bar()
    C-->>A: 
    deactivate A
    end

```
